dialogue-core/src/main/java/com/palantir/dialogue/core/Balanced.java

```diff
@@ -0,0 +1,173 @@
+                    "Channel scores {}",
+                    SafeArg.of("scores", Lists.transform(channels, ChannelWithStats::score)),
+                    SafeArg.of("inflight", Lists.transform(channels, ChannelWithStats::inflight)));
+        }
+
+        // pre-shuffling is pretty important here, otherwise when there are no requests in flight, we'd
+        // *always* prefer the first channel of the list, leading to a higher overall load.
+        List<ChannelWithStats> preShuffled = shuffleImmutableList(channels, random);
+
+        // TODO(dfox): P2C optimization when we have high number of nodes to save CPU?
+        // http://www.eecs.harvard.edu/~michaelm/NEWWORK/postscripts/twosurvey.pdf
+        return preShuffled.stream()
+                .sorted(BY_SCORE)
+                .map(channel -> channel.maybeExecute(endpoint, request))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                // we rely heavily on streams being lazy here
+                .findFirst();
+    }
+
+    /** Returns a new shuffled list, without mutating the input list (which may be immutable). */
+    private static <T> List<T> shuffleImmutableList(ImmutableList<T> sourceList, Random random) {
+        List<T> mutableList = new ArrayList<>(sourceList);
+        Collections.shuffle(mutableList, random);
+        return mutableList;
+    }
+
+    static final class ChannelWithStats implements LimitedChannel {
+        private final LimitedChannel delegate;
+
+        @VisibleForTesting
+        final AtomicInteger inflight = new AtomicInteger(0);
+
+        /**
+         * We keep track of failures within a time window to do well in scenaios where an unhealthy server returns
+         * errors much faster than healthy nodes can serve good responses. See
+         * {@link SimulationTest#fast_503s_then_revert}.
+         */
+        @VisibleForTesting
+        final SlidingTimeWindowArrayReservoir recentFailures;
+
+        // Saves one allocation on each network call
+        private final FutureCallback<Response> updateStats = new FutureCallback<Response>() {
+            @Override
+            public void onSuccess(Response result) {
+                inflight.decrementAndGet();
+                if (Responses.isQosStatus(result) || Responses.isServerError(result)) {
+                    recentFailures.update(1);
+                }
+            }
+
+            @Override
+            public void onFailure(Throwable _throwable) {
+                inflight.decrementAndGet();
+                recentFailures.update(1);
+            }
+        };
+
+        ChannelWithStats(LimitedChannel delegate, CodahaleClock clock) {
+            this.delegate = delegate;
+            this.recentFailures =
+                    new SlidingTimeWindowArrayReservoir(FAILURE_MEMORY.toNanos(), TimeUnit.NANOSECONDS, clock);
+        }
+
+        /** Low = good. */
+        int score() {
+            return inflight.get() + recentFailures.size();
```

Comment:
Perhaps we should scale failures by a factor of ~ten, that way slow successes don't require so many failures to balance out?